_posts/2020-08-14-threshold-tuning/2020-08-14-threshold-tuning.Rmd

```diff
@@ -0,0 +1,227 @@
+against predicting fewer negative cases (second row) or vice versa.
+
+```{r}
+# Lower threshold: More positives
+prd$set_threshold(0.25)$confusion
+```
+
+```{r}
+# Higher threshold: Fewer positives
+prd$set_threshold(0.75)$confusion
+```
+
+This threshold value can now be adjusted optimially for a given measure, such as accuracy. How this can be done is discussed in the following section.
+
+## Adjusting thresholds: Two strategies
+
+Currently `mlr3pipelines` offers two main strategies towards adjusting `classification thresholds`.
+We can either expose the thresholds as a `hyperparameter` of the Learner by using `PipeOpThreshold`.
+This allows us to tune the `thresholds` via an outisde optimizer from `mlr3tuning`.
+
+Alternatively, we can also use `PipeOpTuneThreshold` which automatically tunes the threshold after each learner fit.
+
+In this blog-post, we'll go through both strategies.
+
+## PipeOpThreshold
+
+`PipeOpThreshold` can be put directly after a `Learner`.
+
+A simple example would be:
+
+```{r}
+gr = lrn("classif.rpart", predict_type = "prob") %>>% po("threshold")
+l = GraphLearner$new(gr)
+```
+
+Note, that `predict_type` = "prob" is required for `po("threshold")` to have any effect.
+
+The `thresholds` are now exposed as a `hyperparameter` of the `GraphLearner` we created:
+
+```{r}
+l$param_set
+```
+
+We can now tune those thresholds fom the outside as follows:
+
+Before `tuning`, we have to define which hyperparameters we want to tune over.
+In this example, we only tune over the `thresholds` parameter of the `threshold` pipeop.
+you can easily imagine, that we can also jointly tune over additional hyperparameters, i.e. rpart's `cp` parameter.
+
+As the `Task` we aim to optimize for is a binary task, we can simply specify the threshold param:
+
+```{r}
+library(paradox)
+ps = ParamSet$new(list(
+  ParamDbl$new("threshold.thresholds", lower = 0, upper = 1)
+))
+```
+
+We now create a `AutoTuner`, which automatically tunes the supplied learner over the `ParamSet` we supplied above.
+
+```{r}
+at = AutoTuner$new(
+  learner = l,
+  resampling = rsmp("cv", folds = 3L),
+  measure = msr("classif.ce"),
+  search_space = ps,
+  terminator = trm("evals", n_evals = 5L),
+  tuner = TunerRandomSearch$new()
+)
+
+at$train(tsk("german_credit"))
+```
+
+For multi-class Tasks, this is a little more complicated.
+We have to use a `trafo` to transform a set of `ParamDbl` into the desired format for `threshold.thresholds`:
+A named numeric vector containing the thresholds.
+This can be easily achieved via a `trafo`:
+
+```{r}
+ps = ParamSet$new(list(
+  ParamDbl$new("versicolor", lower = 0, upper = 1),
+  ParamDbl$new("setosa", lower = 0, upper = 1),
+  ParamDbl$new("virginica", lower = 0, upper = 1)
+))
+ps$trafo = function(x, param_set) {
+  list(threshold.thresholds = mlr3misc::map_dbl(x, identity))
+}
+```
```

Comment:
maybe show what the output of the trafo would look like but I guess people can easily follow